ast/desugar/Desugar.cc

```diff
@@ -59,6 +59,12 @@ pair<MethodDef::ARGS_store, InsSeq::STATS_store> desugarArgs(DesugarContext dctx
                 args.emplace_back(node2TreeImpl(dctx, std::move(arg)));
             }
         }
+    } else if (auto *numparams = parser::cast_node<parser::NumParams>(argnode.get())) {
+        // The block use numbered parameters like `_1` or `_9` so we add them as parameters
+        for (int i = 1; i <= numparams->max; i++) {
+            core::NameRef name = dctx.ctx.state.enterNameUTF8("_" + std::to_string(i));
+            args.emplace_back(MK::Local(loc, name));
```

Comment:
So in this case, the loc for any `_1` / `_2` / ... / `_9` in the body will point at the whole block, right? 

For example, here's an example of where this location might show up:

```ruby
# typed: true
[''].each { |x| T.reveal_type(x) }

[''].each { T.reveal_type(_1) }
```

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%5B''%5D.each%20%7B%20%7Cx%7C%20T.reveal_type(x)%20%7D%0A%0A%5B''%5D.each%20%7B%20T.reveal_type(_1)%20%7D)

If I'm not mistaken, I think that on this branch the `From:` for the `_1` will point to the whole line.

Reply:
Ohh good catch. I could use the block location but the error message will remain imprecise.

What I can do is use the numparams stack to store the first location the numbered parameter was encountered. This should pin-point the `from` at the exact place.

Reply:
Ok I pushed a new version that preserve precise locations for the nodes so for:

```rb
bar do
  1 + _2
  puts _1
  puts _2
  T.reveal_type(_2)
  T.reveal_type(_1)
end
```

we get:

```
test/cli/numbered_parameters/numbered_parameters.rb:5: Revealed type: `T.untyped` https://srb.help/7014
    5  |  T.reveal_type(_2)
          ^^^^^^^^^^^^^^^^^
  From:
    test/cli/numbered_parameters/numbered_parameters.rb:18:
    2  |  1 + _2
              ^^

test/cli/numbered_parameters/numbered_parameters.rb:6: Revealed type: `T.untyped` https://srb.help/7014
    6  |  T.reveal_type(_1)
          ^^^^^^^^^^^^^^^^^
  From:
    test/cli/numbered_parameters/numbered_parameters.rb:19:
    3  |  puts _1
               ^^
```